diff --git a/net80211/ieee80211_monitor.c b/net80211/ieee80211_monitor.c
index 3b3f65a..71dac5d 100644
--- a/net80211/ieee80211_monitor.c
+++ b/net80211/ieee80211_monitor.c
@@ -350,7 +353,8 @@ ieee80211_input_monitor(struct ieee80211com *ic, struct sk_buff *skb,
 			 * on the contents of the frame to set pkttype.
 			 */
 			if (tx)
-				pkttype = PACKET_OUTGOING;
+				//pkttype = PACKET_OUTGOING;
+				pkttype = PACKET_OTHERHOST;
 			else if (IEEE80211_IS_MULTICAST(wh->i_addr1)) {
 				if (IEEE80211_ADDR_EQ(wh->i_addr1, dev->broadcast))
 					pkttype = PACKET_BROADCAST;
